runtime/ast/programindices.go

```diff
@@ -0,0 +1,112 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2020 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ast
+
+import (
+	"sync"
+)
+
+// programIndices is a container for all indices of a program's declarations
+//
+type programIndices struct {
+	once sync.Once
+	// Use `pragmaDeclarations` instead
+	_pragmaDeclarations []*PragmaDeclaration
+	// Use `importDeclarations` instead
+	_importDeclarations []*ImportDeclaration
+	// Use `interfaceDeclarations` instead
+	_interfaceDeclarations []*InterfaceDeclaration
+	// Use `compositeDeclarations` instead
+	_compositeDeclarations []*CompositeDeclaration
+	// Use `functionDeclarations()` instead
+	_functionDeclarations []*FunctionDeclaration
+	// Use `transactionDeclarations()` instead
+	_transactionDeclarations []*TransactionDeclaration
+}
+
+func (i *programIndices) pragmaDeclarations(declarations []Declaration) []*PragmaDeclaration {
+	i.once.Do(i.initializer(declarations))
+	return i._pragmaDeclarations
+}
+
+func (i *programIndices) importDeclarations(declarations []Declaration) []*ImportDeclaration {
+	i.once.Do(i.initializer(declarations))
+	return i._importDeclarations
+}
+
+func (i *programIndices) interfaceDeclarations(declarations []Declaration) []*InterfaceDeclaration {
+	i.once.Do(i.initializer(declarations))
+	return i._interfaceDeclarations
+}
+
+func (i *programIndices) compositeDeclarations(declarations []Declaration) []*CompositeDeclaration {
+	i.once.Do(i.initializer(declarations))
+	return i._compositeDeclarations
+}
+
+func (i *programIndices) functionDeclarations(declarations []Declaration) []*FunctionDeclaration {
+	i.once.Do(i.initializer(declarations))
+	return i._functionDeclarations
+}
+
+func (i *programIndices) transactionDeclarations(declarations []Declaration) []*TransactionDeclaration {
+	i.once.Do(i.initializer(declarations))
+	return i._transactionDeclarations
+}
+
+func (i *programIndices) initializer(declarations []Declaration) func() {
+	return func() {
+		i.init(declarations)
+	}
+}
+
+func (i *programIndices) init(declarations []Declaration) {
+
+	// Important: allocate instead of nil
```

Comment:
Why is it important?

Reply:
It used to be important before the refactor, because nil was used as the marker to know if the index had been created or not. The comment is now not necessary anymore.